### PR TITLE
[ENH] Extend column selection to non-strings

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Allow column selection on non-strings, if the columns support it (dates, numbers, ...)
- Allow return of pd.Index where possible
- Allow Pandas to generate error message where possible 
- Move column checks for `index` and `column_names` in `pivot_longer` to `_select_columns`

**This PR relates to #1105 .**

# PR Checklist

Please ensure that you have done the following:

1. [ ] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.md`.

3. [x] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1106--pyjanitor.netlify.app

**Review by ericmjl — APPROVED** [review]

File: janitor/functions/utils.py [hunk 1]
```diff
@@ -211,192 +210,150 @@ def _select_column_names(columns_to_select, df):
...
+    df_columns = df.columns
+    if pd.api.types.is_string_dtype(df_columns):
+        if (
+            "*" in columns_to_select
+        ):  # shell-style glob string (e.g., `*_thing_*`)
+            return fnmatch.filter(df_columns, columns_to_select)
+        if columns_to_select in df_columns:
+            return [columns_to_select]
+        raise KeyError(f"No match was returned for '{columns_to_select}'.")
+    if pd.api.types.is_datetime64_any_dtype(df.columns):
```
Comment (thatlittleboy) [inline janitor/functions/utils.py]: use `df_columns` here?

File: janitor/functions/utils.py [hunk 2]
```diff
@@ -211,192 +210,150 @@ def _select_column_names(columns_to_select, df):
...
-
-        return [*df.columns[columns_to_select]]
-
-    filtered_columns = []
-    columns_to_select = (
-        _select_column_names(entry, df) for entry in columns_to_select
+        try:
+            timestamp = df_columns.get_loc(columns_to_select)
+        except Exception as error:
+            raise error
```
Comment (thatlittleboy) [inline janitor/functions/utils.py]: I'm curious about this pattern. why the need for `try-except` if we aren't doing any special error handling?

File: tests/utils/test__select_column.py [hunk 3]
```diff
@@ -58,16 +77,210 @@ def df_tuple():
...
+    """
+    with pytest.raises(KeyError, match="No match was returned for.+"):
+        _select_column_names("word", df)
+
+
+def test_strings_dates(df_dates):
+    """Test output for datetime column."""
+    assert _select_column_names("2011-01-31", df_dates), df_dates.loc[
+        :, "2011-01-31"
+    ].name
```
Comment (thatlittleboy) [inline tests/utils/test__select_column.py]: I'm don't really understand this test, what exactly are we asserting here? also the assertion error message can be clearer here, I think.

Is it meant to be `assert _select_column_names("2011-01-31", df_dates) == ["2011-01-31"]` or you meant to use `assert_index_equal` maybe?

File: tests/utils/test__select_column.py [hunk 4]
```diff
@@ -307,64 +417,63 @@ def test_callable_string_methods(df1):
...
     assert _select_column_names(["id", "Name"], df1) == ["id", "Name"]
 
 
 def test_list_boolean(df):
     """Test _select_column_names function on list of booleans."""
     booleans = [True, True, True, False, False, False, True, True, True]
-    assert _select_column_names(booleans, df) == list(df.columns[booleans])
+    assert_index_equal(
+        _select_column_names(booleans, df), df.columns[booleans]
+    )
```
Comment (thatlittleboy) [inline tests/utils/test__select_column.py:487]: Is there a test missing for checking an error is raised when the list of booleans' length is not equal to the number of columns?

**Review by thatlittleboy — CHANGES_REQUESTED** [review]

File: janitor/functions/utils.py (hunk 2 above)
Comment (samukweku) [reply janitor/functions/utils.py]: thanks for this ... allowed me to fix a flaw while at it

File: tests/utils/test__select_column.py (hunk 3 above)
Comment (samukweku) [reply tests/utils/test__select_column.py:139]: for string selection on dates, it returns an exact value (pd.Timestamp), assert_index_equal isn't required here, I'm only asserting that two timestamps (which are scalar) are equal

File: tests/utils/test__select_column.py (hunk 4 above)
Comment (samukweku) [reply tests/utils/test__select_column.py:487]: yes ... `test_boolean_list_dtypes`  ... lines 324-336

File: tests/utils/test__select_column.py (hunk 3 above)
Comment (samukweku) [reply tests/utils/test__select_column.py]: fixed the test ... hopefully it's clearer.

Comment (samukweku) [reply tests/utils/test__select_column.py]: thanks for the correction, it is meant to be `assert _select_column_names("2011-01-31", df_dates) == ["2011-01-31"]`

**Review by thatlittleboy — APPROVED** [review]
Thanks for addressing the review comments, the rest LGTM 😄

**Pull request merged**